comp/forwarder/defaultforwarder/blocked_endpoints.go

```diff
@@ -67,15 +121,48 @@ func (e *blockedEndpoints) close(endpoint string) {
 	if knownBlock, ok := e.errorPerEndpoint[endpoint]; ok {
 		b = knownBlock
 	} else {
-		b = &block{}
+		b = &block{
+			endpoint: endpoint,
+			state:    unblocked,
+		}
 	}
 
-	b.nbError = e.backoffPolicy.IncError(b.nbError)
-	b.until = time.Now().Add(e.getBackoffDuration(b.nbError))
+	b.m.RLock()
+
+	switch b.state {
+	case unblocked:
+		// The circuit breaker is unblocked, we need to block for a time determined by the
+		// backoff policy.
+		b.m.RUnlock()
+		b.m.Lock()
+		defer b.m.Unlock()
+
+		if b.state == unblocked {
+			b.nbError = e.backoffPolicy.IncError(b.nbError)
+			b.until = TimeNow().Add(e.getBackoffDuration(b.nbError))
+			b.setState(blocked)
+		}
+	case halfBlocked:
+		// The test transaction failed, so we need to mave back to blocked.
+		b.m.RUnlock()
+		b.m.Lock()
+		defer b.m.Unlock()
+
+		if b.state == halfBlocked {
+			b.nbError = e.backoffPolicy.IncError(b.nbError)
+			b.until = TimeNow().Add(e.getBackoffDuration(b.nbError))
+			b.setState(blocked)
+		}
```

Comment:
This code is identical between `blocked` and `halfBlocked` branches. Perhaps this switch could be an `if` statement insteead? Like

```
if b.state != blocked {
   ...
}